### PR TITLE
core: Allow refresh of local state with no resources

### LINKED ISSUE #4154: Local and remote state conflict on no changes
Hi,

I have stumbled upon issue with remote states. We use these for different TF deployments. Whereas usually it works, in one simple case it breaks. Steps to reproduce:

1. `terraform apply -var env=test`
2. `terraform remote config -backend=s3 -backend-config="bucket=test-state" -backend-config="key=bucket.tfstate"` 
3. Go to a different system/container and check out terraform files
4. `terraform remote config -backend=s3 -backend-config="bucket=test-state" -backend-config="key=bucket.tfstate"`

This results in
```
Remote configuration updated
Error while performing the initial pull. The error message is shown
below. Note that remote state was properly configured, so you don't
need to reconfigure. You can now use `push` and `pull` directly.

Unknown refresh result: Local and remote state conflict, manual resolution required
```

The local .terraform/terraform.tfstate file is
```
{
    "version": 1,
    "serial": 0,
    "remote": {
        "type": "s3",
        "config": {
            "bucket": "test-state",
            "key": "bucket.tfstate"
        }
    },
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {}
        }
    ]
}
```

The remote bucket.tfstate file is
```
{
    "version": 1,
    "serial": 0,
    "remote": {
        "type": "s3",
        "config": {
            "bucket": "test-state",
            "key": "bucket.tfstate"
        }
    },
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {
                "bucket": "test-state",
                "environment": "test",
                "region": "eu-west-1"
            },
            "resources": {
                "aws_s3_bucket.terraform-state-s3": {
                    "type": "aws_s3_bucket",
                    "primary": {
                        "id": "test-state",
                        "attributes": {
                            "acl": "private",
                            "arn": "arn:aws:s3:::test-state",
                            "bucket": "test-state",
                            "cors_rule.#": "0",
                            "force_destroy": "false",
                            "hosted_zone_id": "Z1BKCTXD74EZPE",
                            "id": "test-state",
                            "policy": "",
                            "region": "eu-west-1",
                            "tags.#": "0",
                            "versioning.#": "1",
                            "versioning.69840937.enabled": "true",
                            "website.#": "0"
                        }
                    }
                }
            }
        }
    ]
}
```

The terraform files:
bucket.tf
```
provider "aws" {
  region = "${var.region}"
}

resource "aws_s3_bucket" "terraform-state-s3" {
  bucket = "${var.env}-state"
  acl = "private"
  force_destroy = "false"
  versioning {
    enabled = true
  }
}
```

outputs.tf

```
output "environment" {
  value = "${var.env}"
}

output "region" {
  value = "${var.region}"
}

output "bucket" {
  value = "${aws_s3_bucket.terraform-state-s3.id}"
}
```
variables.tf
```
variable "env" {
  description = "Environment name"
}

variable "region" {
  description = "AWS region"
  default     = "eu-west-1"
}

```
I am not sure what is TF trying to merge. There is no local state before fetching remote. What's even more curious, I can fetch remote state from unrelated TF build OK (a build which has completely different terraform files). I suspect the issue is somewhere in the remote `bucket.tfstate` file itself...

### LINKED ISSUE #5410: Re-configuring remote state raises conflict on pull
I'm currently working through a POC using CircleCI to manage plans & applies of Terraform state, using S3 as the remote backend. Circle won't be keeping the filesystem around, so relying on the `.terraform` folders locally to stick around isn't possible. When doing a plan (when a PR is opened), I run the following commands:

1. `terraform remote config [...]`
2. `terraform remote pull`
3. `terraform plan -out plan.out`

First pass through worked fine and I could apply state, but if I test the condition where .terraform does not exist, I'm getting unexpected behaviors:

```
terraform remote config -backend=s3 -backend-config="bucket=foo" -backend-config="key=states/aws/s3_preprod_useast1/terraform.tfstate" -backend-config="region=us-east-1"
Initialized blank state with remote state enabled!
Error while performing the initial pull. The error message is shown
below. Note that remote state was properly configured, so you don't
need to reconfigure. You can now use `push` and `pull` directly.

Unknown refresh result: Local and remote state conflict, manual resolution required
```

For giggles, I run pull just to make sure I'm not crazy:

```
$ terraform remote pull
Failed to read state: Unknown refresh result: Local and remote state conflict, manual resolution required
```

The local state looks like this:

```
# .terraform/terraform.tfstate
{
    "version": 1,
    "serial": 0,
    "remote": {
        "type": "s3",
        "config": {
            "bucket": "foo",
            "key": "states/aws/s3_preprod_useast1/terraform.tfstate",
            "region": "us-east-1"
        }
    },
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {}
        }
    ]
}
```

And the remote state in s3 looks like this:

```
# s3://foo/states/aws/s3_preprod_useast1/terraform.tfstate
{
    "version": 1,
    "serial": 0,
    "remote": {
        "type": "s3",
        "config": {
            "bucket": "foo",
            "key": "states/aws/s3_preprod_useast1/terraform.tfstate",
            "region": "us-east-1"
        }
    },
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {
                "aws_s3_bucket.rob_test": {
                    "type": "aws_s3_bucket",
                    "primary": {
                        "id": "foo-rob-test",
                        "attributes": {
                            "acl": "private",
                            "arn": "arn:aws:s3:::foo-rob-test",
                            "bucket": "foo-rob-test",
                            "cors_rule.#": "0",
                            "force_destroy": "false",
                            "hosted_zone_id": "xxxx",
                            "id": "foo-rob-test",
                            "policy": "{\"Statement\":[{\"Action\":\"s3:GetObject\",\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"*\"},\"Resource\":\"arn:aws:s3:::foo-rob-test/*\",\"Sid\":\"AddPerm\"}],\"Version\":\"2008-10-17\"}",
                            "region": "us-east-1",
                            "tags.#": "0",
                            "website.#": "0"
                        }
                    }
                }
            }
        }
    ]
}
```

I would have expected that `remote config` would've actually pulled the state, but it doesn't. For extra sanity, I also tried adding "-pull=true" while running `remote config`, just in case there was a documentation breakdown.

I verified that everything works fine if I just pull the remote state:

```
$ aws s3 cp s3://foo/states/aws/s3_preprod_useast1/terraform.tfstate terraform.tfstate --region us-east-1
$ mv terraform.tfstate .terraform/terraform.tfstate
$ terraform remote pull
Local and remote state in sync
```

I'm on 0.6.12, but it was happening on 0.6.11 as well.

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes #4154 and fixes #5410.

`HasResources` is copied from @apparentlymart's work in #7026.

### CODE REVIEW

**Comment by dtolnay** [comment]
I assume this is waiting until after the 0.7 release

**Comment by dtolnay** [comment]
@jen20 Is there anything I can do to move this along?

File: state/cache.go [hunk 1]
```diff
@@ -64,8 +64,13 @@ func (s *CacheState) RefreshState() error {
 		// Cache is newer than remote. Not a big deal, user can just
 		// persist to get correct state.
 		s.refreshResult = CacheRefreshLocalNewer
-	case cached == nil && durable != nil:
+	case !cached.HasResources() && durable != nil:
```
Comment (jen20) [inline state/cache.go]: We should probably verify that cached is not nil here since the `HasResources()` method accesses members.

**Comment by jen20** [comment]
HI @dtolnay! Looks like we have a problem with the test applied to master:

```
--- FAIL: TestCacheState_RefreshState (0.01s)
panic: missing module path [recovered]
       	panic: missing module path

goroutine 39 [running]:
panic(0x345f40, 0xc42021e050)
       	/usr/local/Cellar/go/1.7/libexec/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc4201d4840)
       	/usr/local/Cellar/go/1.7/libexec/src/testing/testing.go:579 +0x25d
panic(0x345f40, 0xc42021e050)
       	/usr/local/Cellar/go/1.7/libexec/src/runtime/panic.go:458 +0x243
github.com/hashicorp/terraform/terraform.(*State).ModuleByPath(0xc420216960, 0x56bd10, 0x1, 0x1, 0xc420090838)
       	/Users/James/Code/go/src/github.com/hashicorp/terraform/terraform/state.go:145 +0x1e2
github.com/hashicorp/terraform/terraform.(*State).init(0xc420216960)
       	/Users/James/Code/go/src/github.com/hashicorp/terraform/terraform/state.go:580 +0x62
github.com/hashicorp/terraform/terraform.WriteState(0xc420216960, 0x553a40, 0xc420090a78, 0x0, 0x0)
       	/Users/James/Code/go/src/github.com/hashicorp/terraform/terraform/state.go:1775 +0x51
github.com/hashicorp/terraform/state.(*LocalState).WriteState(0xc4201f1380, 0xc420216960, 0x0, 0x0)
       	/Users/James/Code/go/src/github.com/hashicorp/terraform/state/local.go:69 +0x19c
github.com/hashicorp/terraform/state.(*CacheState).WriteState(0xc420051ec8, 0xc420216960, 0xc4201f1600, 0x3c)
       	/Users/James/Code/go/src/github.com/hashicorp/terraform/state/cache.go:28 +0x41
github.com/hashicorp/terraform/state.TestCacheState_RefreshState(0xc4201d4840)
       	/Users/James/Code/go/src/github.com/hashicorp/terraform/state/cache_test.go:91 +0x3f2
testing.tRunner(0xc4201d4840, 0x3fb5f8)
       	/usr/local/Cellar/go/1.7/libexec/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
       	/usr/local/Cellar/go/1.7/libexec/src/testing/testing.go:646 +0x2ec
FAIL   	github.com/hashicorp/terraform/state   	0.044s
```

One of us will pick this up next week unless you get a chance to do so beforehand. Thanks!

File: state/cache.go (hunk 1 above)
Comment (dtolnay) [inline state/cache.go]: `HasResources` returns false if the state is nil.

**Comment by dtolnay** [comment]
I'm surprised it still merges cleanly after 2.5 months. I will let you folks take care of the test failure. I enabled the brand new ["allow edits from maintainers"](https://github.com/blog/2247-improving-collaboration-with-forks) to make it easier.

**Comment by dtolnay** [comment]
Never mind, it was simple enough I fixed it myself.

**Comment by dtolnay** [comment]
`git bisect` points to #8120 as the point it started failing.

**Comment by zachgersh** [comment]
I would love to +1 this (can it be merged soonish) - just needed this today and thought terraform would work this way when it clearly does not.

**Comment by dtolnay** [comment]
@jen20 is there anything else I can do to help you merge this?

**Pull request merged**